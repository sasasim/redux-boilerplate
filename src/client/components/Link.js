import React, { Component, PropTypes } from 'react';

export default class Link extends Component {
  onClick = (event) => {
    if (!event.button && !event.altKey && !event.ctrlKey && !event.metaKey && !event.shiftKey) {
      event.preventDefault();
      const { name, params, options } = this.props;
      this.context.router.navigate(name, params, options);
    }
  }

  render() {
    const { children, name, params } = this.props;
    const { router } = this.context;
    const active = router.isActive(name, params);
    const href = router.buildUrl(name, params);
    return (
      <a
        className={active ? 'active-link' : null}
        href={href}
        onClick={this.onClick}
      >{children}</a>
    );
  }
}

Link.propTypes = {
  children: PropTypes.node.isRequired,
  name: PropTypes.string.isRequired,
  options: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  params: PropTypes.object // eslint-disable-line react/forbid-prop-types
};

Link.contextTypes = {
  router: PropTypes.object.isRequired
};
