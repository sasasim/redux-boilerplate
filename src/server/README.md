# Node.js server

## Rendering

Boilerplate server is shipped with two options of how to render the content.

1. `src/server/rendering/plainRendering` statically renders the HTML template which is defined in `src/server/rendering/template.js` it does not render any content in `root` div, so it's basically used for client-side rendered only applications.

2. `src/server/rendering/serverSideRendering` renders React application (implemented in client) on the server, it also respects routing, HTTP status codes etc. Just like traditional non SPA application. When the page is rendered and loaded into the browser, React will kick in, attaches all the event handlers and from now on the application is traditional SPA.

Please, keep in mind that when SSR is not a requirement, it's not reccomended to use `serverSideRendering` module, because it simply makes things harder, we can't guarantee that all the code can be used altogether on the server and client. Server side rendering is also an operation which is not for free, theoretically in huge number of requests / second, this overhead is non marginal.


