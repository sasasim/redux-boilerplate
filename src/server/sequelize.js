import Sequelize from 'sequelize';

export default new Sequelize(process.env.PSQL);
