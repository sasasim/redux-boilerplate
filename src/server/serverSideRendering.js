import React from 'react';
import { renderToString } from 'react-dom/server';
import { Provider } from 'react-redux';
import { RouterProvider } from 'react-router5';

import buildRouter from 'src/client/router';
import buildStore from 'src/client/store';
import Root from 'src/client/components/Root';

import stats from './stats.json';

const template = (content, js, css) => `
  <!doctype html>
  <html>
    <head>
      <meta charset="utf-8">
      <link href="${css}" rel="stylesheet">
    </head>
    <body>
      <div id="root">${content}</div>
      <script type="text/javascript" src="${js}"></script>
    </body>
  </html>
`;

export default (req, res, next) => {
  const router = buildRouter();
  const store = buildStore(router);

  router.start(req.originalUrl, (error) => {
    const { router: { route } } = store.getState();

    if (route) {
      res
        .status(200)
        .send(template(renderToString(
          <Provider store={store}>
            <RouterProvider router={router}>
              <Root />
            </RouterProvider>
          </Provider>
        ), stats.js, stats.css));
    } else {
      next();
    }
  });
};
