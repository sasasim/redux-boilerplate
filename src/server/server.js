import express from 'express';
import compression from 'compression';

import apiRouter from 'src/server/api/router';
import serverSideRendering from 'src/server/rendering/serverSideRendering';
import sequelize from 'src/server/sequelize';

sequelize
  .authenticate()
  .then(() => console.log('Connection has been established successfully.'))
  .catch(err => console.error('Unable to connect to database.', err));

const server = express();

// gzip compression
server.use(compression());

// serving static files
server.use('/', express.static('./dist/client'));

// API router, for serving the API
server.use('/api/v1', apiRouter);

// Choose appropriate rendering based on your
// use case
server.use('*', serverSideRendering);

server.listen(process.env.PORT || 3001);
