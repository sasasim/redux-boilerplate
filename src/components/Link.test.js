import React, { Component, PropTypes } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import Link from 'src/components/Link';

class RouterProvider extends Component {
  getChildContext() {
    return { router: this.router };
  }

  router = {
    buildUrl: (name, params) => {
      this.props.log.push(['buildUrl', name, params]);
      return `test/${name}`;
    },
    isActive: (url) => {
      this.props.log.push(['isActive', url]);
      return url === 'test-route2';
    }
  };

  render() {
    return this.props.children;
  }
}

RouterProvider.childContextTypes = {
  router: PropTypes.object
};

RouterProvider.propTypes = {
  children: PropTypes.node.isRequired,
  log: PropTypes.array.isRequired // eslint-disable-line react/forbid-prop-types
};

describe('Link', () => {
  it('should render inactive', () => {
    const log = [];
    log.push(['result', renderToStaticMarkup(
      <RouterProvider log={log}>
        <Link name="test-route1">
          <test-child />
        </Link>
      </RouterProvider>)]);
    expect(log).toMatchSnapshot();
  });

  it('should render active', () => {
    const log = [];
    log.push(['result', renderToStaticMarkup(
      <RouterProvider log={log}>
        <Link name="test-route2">
          <test-child />
        </Link>
      </RouterProvider>)]);
    expect(log).toMatchSnapshot();
  });

  it('should render with params', () => {
    const log = [];
    const testParams = { testParam1: 'testValue1', testParam2: 'testValue2' };
    log.push(['result', renderToStaticMarkup(
      <RouterProvider log={log}>
        <Link name="test-route1" params={testParams}>
          <test-child />
        </Link>
      </RouterProvider>)]);
    expect(log).toMatchSnapshot();
  });
});
