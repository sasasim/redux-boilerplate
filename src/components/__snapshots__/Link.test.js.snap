exports[`Link should render active 1`] = `
Array [
  Array [
    "isActive",
    "test-route2",
  ],
  Array [
    "buildUrl",
    "test-route2",
    undefined,
  ],
  Array [
    "result",
    "<a class=\"active-link\" href=\"test/test-route2\"><test-child></test-child></a>",
  ],
]
`;

exports[`Link should render inactive 1`] = `
Array [
  Array [
    "isActive",
    "test-route1",
  ],
  Array [
    "buildUrl",
    "test-route1",
    undefined,
  ],
  Array [
    "result",
    "<a href=\"test/test-route1\"><test-child></test-child></a>",
  ],
]
`;

exports[`Link should render with params 1`] = `
Array [
  Array [
    "isActive",
    "test-route1",
  ],
  Array [
    "buildUrl",
    "test-route1",
    Object {
      "testParam1": "testValue1",
      "testParam2": "testValue2",
    },
  ],
  Array [
    "result",
    "<a href=\"test/test-route1\"><test-child></test-child></a>",
  ],
]
`;
